### ISSUE TITLE
Lazy loading for NER-task (or generically)

### ISSUE THREAD

**Issue by plc-dev** [ThilinaRajapakse/simpletransformers#514]
**Is your feature request related to a problem? Please describe.**
As of right now, the NERModel doesn't support lazy loading for conll files which do not fit into memory.
Also the behaviour of data loading is not consistent between tasks, as the ClassificationModel does support lazy loading.

**Describe the solution you'd like**
I would propose a generic approach to abstract the data loading into it's own class, to have a uniform set of features for all models.
This might conflict with the "simple" approach if a new Class for the dataset has to be instantiated by the user, but the instantiation of the proposed data-loading class could also be handled internally without changing the current interface and have the behaviour configured via additional parameters in the args-dictionary.

**Additional context**
This should be doable by creating subclasses of the PyTorch DataSet-Class which is then handled by the PyTorch DataLoader-Class to iteratively access the data. Further information for an example implementation can be found [here](https://discuss.pytorch.org/t/loading-huge-data-functionality/346/3) and [here](https://discuss.pytorch.org/t/dataloaders-multiple-files-and-multiple-rows-per-column-with-lazy-evaluation/11769/9).

**Comment by ThilinaRajapakse** [comment]
I don't think it would be possible to unify the data loading logic for all tasks. Even if it were possible, I'm not sure it would be a good idea. The _intuitive_ input format is heavily dependent on each task, and I worry that attempting to unify the formats would result in something that is not intuitive for _any_ task. I'm also concerned that it would make maintaining the different tasks and adding new tasks/models far more complicated.

I agree that lazy loading support would be beneficial in all tasks, but I feel that it's better to keep the exact implementation separate and independent of each other.

**Comment by Pradhy729** [comment]
I can help out with creating a lazy loading for NER and others that don't have it.

**Comment by ThilinaRajapakse** [comment]
Thank you, @Pradhy729 , I highly appreciate your offer!

Let's start with NER. The first step is to figure out a good input format. It's easy for classification as you can just have the samples in a text file where each row corresponds to one sample. Unfortunately, with NER, a single sample can be spread out over multiple lines and the number of lines per sample is not necessarily constant for different samples.

The simplest solution I can think of right now is to require each sample/sentence to be in its own text file. Then, we can have the Pytorch dataset read in each file on the fly.

There are a couple of things I don't like about this though. First one is that it introduces an entirely new input format which may be confusing and annoying for users. The second is that this could lead to millions of files in a single directory which could have performance impacts in some OSes/file systems

A tentative solution to both these problems is to handle the conversion from the current input format into the "lazy format" internally. This way, the user doesn't need to learn a new format nor would they have to worry about a directory with tons of files in it.

If love to hear what you guys think!

**Comment by plc-dev** [comment]
I see your point on unifying the the data loading logic. 
The task specific shape of the input eventually makes a difference on how the input has to be streamed, thus is relevant to each models implemenation. 

But I'm a little confused about the idea of the intuitive input format.
As of right now passing a dataframe has the same effect for each task, but passing a file with the task specific shape results in different behaviour. Lazy loading for the ClassificationModel and loading the data entirely into memory for the NERModel. Also the way the data is passed shouldn't determine wether it is kept in memory or loaded lazily and makes it _less_ intuitive imo. Especially if the behaviour isn't uniform across the framework. But I might be missing something here, or misunderstanding your reference of an "intuitive format".

Regarding the input format for the NER task. I wouldn't change the conll format at all, since it is already streamable. 
Each datapoint can be read seperately by reading in lines until a line solely containing a "\n"-character is reached. If two consecutive "\n" are read the file is completely read. 
I would avoid an additional input format, or even splitting into individual files, for the reasons you just mentioned and rather enable the lazy loading by passing an additional parameter, which could make the interface uniform across all tasks again.

**Comment by ThilinaRajapakse** [comment]
Embarrassingly, I had the input formats for my own library mixed up in my head. :sweat_smile: 

I was under the impression that the lazy loading was triggered by a parameter from the args-dict. I do agree that the current flow of having the input file trigger lazy loading is unintuitive, especially considering some models support input through files _without_ lazy loading. This is not the intended behavior and simply a mistake on my part.

The reference to the intuitive input format is with regard to the structure of the Dataframe for each task (`prefix`, `input_text`, `target_text` for `T5Model`, and `text`, `labels` for `ClassificationModel` etc.). I wholeheartedly agree that passing in the data as a DF or as a file path should not result in different behavior. I have no idea what I was thinking. :man_shrugging: 

The intended behavior is that lazy loading should be triggered by the parameter, but lazy loading comes with a _constraint_ that requires input as a file. Once lazy loading is implemented for all tasks, this would mean that the interface is uniform again as you point out.

Agreed again regarding the NER input format. We can use the blank line logic to build up a dictionary that maps all the individual sequences to their line number range in the file. This dictionary can then be used to read the relevant lines from the file for any "random" sequence as required for the train dataloader.

What do you think?

Again, I am sorry about the confusion! :sweat_smile:

**Comment by Pradhy729** [comment]
Yeah the blank lines logic with a dict mapping example numbers to line numbers was my idea as well. I am working on it now - will share when I have a working version.

**Comment by Pradhy729** [comment]
Hi @ThilinaRajapakse  - please see my implementation of Lazy loader dataset for NER: 
1. [New dataset subclass](https://github.com/Pradhy729/simpletransformers/blob/ner_lazy_dataloading/simpletransformers/ner/ner_utils.py#L277)  
2. [PR in my fork](https://github.com/Pradhy729/simpletransformers/pull/7)

Only additional thing that requires change in the main model class is to set a data_start_line in args. This is just to tell my loader at what line number the first example starts. This may depend on the input file format. Let me know if you have any comments or concerns and I can create a PR.

**Comment by ThilinaRajapakse** [comment]
Looks good from what I can see. Please submit the PR and I'll test and merge. Thank you!
